dockerfiles/base/scripts/base/library.sh

```diff
@@ -517,3 +524,92 @@ wait_until_server_is_booted() {
     ELAPSED=$((ELAPSED+1))
   done
 }
+
+# Compares $1 version to the first 10 versions listed as tags on Docker Hub
+# Returns "" if $1 is newest, otherwise returns the newest version available
+compare_versions() {
+
+  local VERSION_LIST_JSON=$(curl -s https://hub.docker.com/v2/repositories/${CHE_IMAGE_NAME}/tags/)
+  local NUMBER_OF_VERSIONS=$(echo $VERSION_LIST_JSON | jq '.count')
+
+  DISPLAY_LIMIT=10
+  if [ $DISPLAY_LIMIT -gt $NUMBER_OF_VERSIONS ]; then 
+    DISPLAY_LIMIT=$NUMBER_OF_VERSIONS
+  fi
+
+  # Strips off -M#, -latest version information
+  BASE_VERSION=$(echo $1 | cut -f1 -d"-")
+
+  COUNTER=0
+  RETURN_VERSION=""
+  while [ $COUNTER -lt $DISPLAY_LIMIT ]; do
+    TAG=$(echo $VERSION_LIST_JSON | jq ".results[$COUNTER].name")
+    TAG=${TAG//\"}
+
+    if [ "$TAG" != "nightly" ] && [ "$TAG" != "latest" ]; then
+      if less_than $BASE_VERSION $TAG; then
+        RETURN_VERSION=$TAG
+        break;
+      fi
+    fi
+    let COUNTER=COUNTER+1 
+  done
+
+  echo $RETURN_VERSION
+}
+
+# Input - an array of ports and port descriptions to check
+# Output - true if all ports are open, false if any of them are already bound
+check_all_ports(){
+
+  declare -a PORT_INTERNAL_ARRAY=("${@}")
+
+  DOCKER_PORT_STRING=""
+  HTTPD_PORT_STRING=""
+  for index in "${!PORT_INTERNAL_ARRAY[@]}"; do 
+    PORT=${PORT_INTERNAL_ARRAY[$index]%;*}
+    PORT_STRING=${PORT_INTERNAL_ARRAY[$index]#*;}
+
+    DOCKER_PORT_STRING+=" -p $PORT:$PORT"
+    HTTPD_PORT_STRING+=" -p $PORT"
+  done
+
+  # Need to evaluate a string otherwise the ports will be seen as strings and not parameters
+  EXECUTION_STRING="docker run -d ${DOCKER_PORT_STRING} \
+                      --name fake ${UTILITY_IMAGE_ALPINE} \
+                          httpd -f ${HTTPD_PORT_STRING} -h /etc/ > /dev/null 2>&1"
+
+  eval ${EXECUTION_STRING}
+  NETSTAT_EXIT=$?
+
+  docker rm -f fake > /dev/null 2>&1
```

Comment:
I've tried on ubuntu and mac and I don't need to use httpd to fail if port mapping is already in use

```
benoitf@ubuntu:~/che/dockerfiles$ docker run -p 22:22  -ti alpine:3.4 sh -c "echo hello"
docker: Error response from daemon: driver failed programming external connectivity on endpoint angry_hawking (a1ffa937fac9bf49ddfa8bfd1cc42199c887b7e4647cc48e628015d3216c1533): Error starting userland proxy: listen tcp 0.0.0.0:22: bind: address already in use.
```
but well it's not a big issue

another item : why do we need to launch container as a daemon and then remove it later.
it is not working on a single command (no daemon/background) ?